bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieClientImpl.java

```diff
@@ -325,11 +339,33 @@ public void addEntry(final BookieId addr,
         // Retain the buffer, since the connection could be obtained after
         // the PendingApp might have already failed
         toSend.retain();
-
+        Optional<WriteAndFlushCallback> callback = Optional.empty();
+        try {
+            callback = setMemoryLimit(entryId, toSend.readableBytes());
```

Comment:
It seems to me that here it is to late to do the memory accounting as the buffer is already allocated we should move this before the allocation of the buffer that is going to be retained.
Also, we should really make it clear which is the memory that we are counting with this counter.